### ISSUE TITLE
Exclusions don't work when classifiers are involved

### ISSUE THREAD

**Issue by trenchguinea** [spring-gradle-plugins/dependency-management-plugin#80]
From the issue summary this might sound similar to #67, but this is different.  I have found that when an excluded module has a Maven classifier in its artifact metadata it does not get excluded when defined in dependencyManagement but does get excluded when defined in the Gradle core dependencies exclude.  Here is an example that replicates the issue.

First, let's start with this build script that simply defines 1 dependency and pulls in avro-mapred-1.7.7-hadoop2.jar:

    plugins {
      id 'io.spring.dependency-management' version '0.5.6.RELEASE'
    }

    apply plugin: 'java'
    apply plugin: 'io.spring.dependency-management'

    repositories {
      mavenLocal()
      mavenCentral()
      maven { url "https://repository.cloudera.com/artifactory/cloudera-repos/" }
    }

    dependencies {
      compile(group: 'org.apache.crunch', name: 'crunch-core', version: '0.8.4+3mat-cdh4.6.0') {
        exclude module: 'parquet-avro'
        exclude module: 'trevni-avro'
        exclude module: 'trevni-core'
      }
    }

    task runtimeArtifacts << {
      def sortedArtifacts = new TreeSet()
      configurations.runtime.each { sortedArtifacts.add(it.name) }
      sortedArtifacts.each { println it }  
    }

`gradle runtimeArtifacts` will show `avro-mapred-1.7.7-hadoop2.jar`.

Now, I'll add another dependency that pulls in avro-mapred but this time with the "hadoop1" classifier.

    dependencies {
      compile(group: 'org.apache.crunch', name: 'crunch-core', version: '0.8.4+3mat-cdh4.6.0') {
        exclude module: 'parquet-avro'
        exclude module: 'trevni-avro'
        exclude module: 'trevni-core'
      }
      compile(group: 'org.kitesdk', name: 'kite-hadoop1-dependencies', version: '1.1.0')
    }

`gradle runtimeArtifacts` will now show both `avro-mapred-1.7.7-hadoop1.jar` and `avro-mapred-1.7.7-hadoop2.jar`.

Because this is invalid I can exclude avro-mapred from kite-hadoop1-dependencies, and it will work:

    dependencies {
      compile(group: 'org.apache.crunch', name: 'crunch-core', version: '0.8.4+3mat-cdh4.6.0') {
        exclude module: 'parquet-avro'
        exclude module: 'trevni-avro'
        exclude module: 'trevni-core'
      }
      compile(group: 'org.kitesdk', name: 'kite-hadoop1-dependencies', version: '1.1.0') {
        exclude module: 'avro-mapred'
      }
    }

`gradle runtimeArtifacts` will correctly show only `avro-mapred-1.7.7-hadoop2.jar`.

However, if I use dependencyManagement to exclude this avro-mapred module then it will not work:

    dependencyManagement {
      dependencies {
        dependency(group: 'org.kitesdk', name: 'kite-hadoop1-dependencies', version: '1.1.0') {
          exclude group: 'org.apache.avro', name: 'avro-mapred'
        }
      }
    }

    dependencies {
      compile(group: 'org.apache.crunch', name: 'crunch-core', version: '0.8.4+3mat-cdh4.6.0') {
        exclude module: 'parquet-avro'
        exclude module: 'trevni-avro'
        exclude module: 'trevni-core'
      }
      compile group: 'org.kitesdk', name: 'kite-hadoop1-dependencies'
    }

`gradle runtimeArtifacts` will show both `avro-mapred-1.7.7-hadoop1.jar` and `avro-mapred-1.7.7-hadoop2.jar`.

In summary, I am not able to globally manage the exclusion of dependencies that use classifiers.

**Comment by trenchguinea** [comment]
Actually, my issue description is a little too broad.  It's not that modules with classifiers never work, it's just that they don't work when a different dependency pulls in a different classifier version.  For example, if I remove that dependency on crunch-core then the avro-mapred module will be correctly excluded.

**Comment by wilkinsona** [comment]
Thanks for the detailed report. Unfortunately, initial impressions are that there's nothing I can do about this.

Like in #67, information about classifiers isn't available when walking a dependency tree from `configuration.incoming.resolutionResult.root`. That means that the two `org.apache.avro:avro-mapred` dependencies look exactly the same from this plugin's perspective. It sees that the dependency is excluded beneath `org.kitesdk:kite-hadoop1-dependencies`, but isn't excluded beneath `org.apache.crunch:crunch-core` which means that the dependency is retained.

To fix this, there'd need to be a way to figure out a dependency's classifier so that the plugin can correctly differentiate between the two `avro-mapred` dependencies and correctly exclude one of them.

**Comment by wilkinsona** [comment]
One possibility: unlike #67, when processing exclusions a dependencies' pom files are available. The pom files could be used as a source for the classifiers.

**Comment by wilkinsona** [comment]
I've written a [post on Gradle's forums](https://discuss.gradle.org/t/getting-information-about-a-dependencys-classifier/16685) to see if there's a better way to get the classifier. Without an official API, I think this may have to remain as a known limitation as I'm really not sure that figuring out the classifier by looking in the pom is a good idea.

**Comment by trenchguinea** [comment]
Thanks, Andy.  What's interesting to me is that the module-level exclusion is able to recognize that the dependency version is different and exclude it from `kite-hadoop1-dependencies` while keeping it in `crunch-core`.  At least some parts of Gradle are able to understand classifiers, just not all of it.

**Comment by wilkinsona** [comment]
I don't think any understand of classifiers is required at that point. It's simply excluding any `avro-mapred` dependency that's beneath `kite-hadoop1-dependencies`. As the other `avro-mapred` dependency isn't beneath `kite-hadoop1-dependencies` it isn't excluded.

Gradle actually recommends that exclusions are applied to an entire configuration. At that point, the fact that it doesn't understand classifiers would be problematic. This fails as there's no `classifier` property on `DefaultExcludeRule`:

```
configurations {
  compile.exclude module: 'avro-mapred', classifier: 'hadoop1'
}
```

And this excludes both `avro-mapred` dependencies:

```
configurations {
  compile.exclude module: 'avro-mapred', classifier: 'hadoop1'
}
```

**Comment by trenchguinea** [comment]
You're right.  I was thinking of a more complicated situation where module-level exclusions don't always work and was thinking that in this situation it did work because of classifier recognition, but with more testing I found that the module-level exclusion is working simply because of the simplicity of this contrived example.

Btw, did you mean for your code sample to be exactly the same in your examples in the prior comment?

**Comment by wilkinsona** [comment]
I didn't. Fixed now.

**Comment by wilkinsona** [comment]
Sadly, there doesn't appear to be any interest on the Gradle side to provide the required API making this impossible to implement elegantly. With some regret, I'm going to close this issue with an update to the README explaining the limitation
